File: DataStructure/DesignSinglyLinkedList.cpp
#include <vector>
using namespace std;

// Node class for the linked list
class Node {
public:
    int data;
    Node* next;
    Node(int val) : data(val), next(nullptr) {}
};

class LinkedList {
private:
    Node* head; // Pointer to the head of the linked list

public:
    // Constructor to initialize an empty linked list
    LinkedList() : head(nullptr) {}

    // Function to get the value of the ith node (0-indexed)
    // Returns -1 if index is out of bounds
    int get(int index) {
        if (index < 0)
            return -1;

        Node* current = head;
        int count = 0;
        while (current != nullptr && count < index) {
            current = current->next;
            count++;
        }
        if (current != nullptr)
            return current->data;
        else
            return -1;
    }

    // Function to insert a node with value val at the head of the list
    void insertHead(int val) {
        Node* newNode = new Node(val);
        newNode->next = head;
        head = newNode;
    }
    
    // Function to insert a node with value val at the tail of the list
    void insertTail(int val) {
        Node* newNode = new Node(val);
        if (head == nullptr) {
            head = newNode;
            return;
        }

        Node* current = head;
        while (current->next != nullptr) {
            current = current->next;
        }
        current->next = newNode;
    }

    // Function to remove the ith node (0-indexed)
    // Returns false if index is out of bounds, otherwise returns true
    bool remove(int index) {
        if (index < 0)
            return false;
        
        if (index == 0) {
            if (head != nullptr) {
                Node* temp = head;
                head = head->next;
                delete temp;
                return true;
            } else {
                return false;
            }
        }

        Node* current = head;
        int count = 0;
        while (current != nullptr && count < index - 1) {
            current = current->next;
            count++;
        }

        if (current == nullptr || current->next == nullptr)
            return false;

        Node* temp = current->next;
        current->next = temp->next;
        delete temp;
        return true;
    }

    // Function to return an array of all the values in the linked list, ordered from head to tail
    vector<int> getValues() {
        vector<int> values;
        Node* current = head;
        while (current != nullptr) {
            values.push_back(current->data);
            current = current->next;
        }
        return values;
    }
};

File: DataStructure/DesignSinglyLinkedList.js
class Node {
    constructor(val) {
        this.data = val;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }

    /**
     * @param {number} index
     * @return {number}
     */
    get(index) {
        if (index < 0 || this.head === null)
            return -1;

        let current = this.head;
        let count = 0;
        while (current !== null && count < index) {
            current = current.next;
            count++;
        }
        if (current !== null)
            return current.data;
        else
            return -1;
    }

    /**
     * @param {number} val
     * @return {void}
     */
    insertHead(val) {
        let newNode = new Node(val);
        newNode.next = this.head;
        this.head = newNode;
    }

    /**
     * @param {number} val
     * @return {void}
     */
    insertTail(val) {
        let newNode = new Node(val);
        if (this.head === null) {
            this.head = newNode;
            return;
        }

        let current = this.head;
        while (current.next !== null) {
            current = current.next;
        }
        current.next = newNode;
    }

    /**
     * @param {number} index
     * @return {boolean}
     */
    remove(index) {
        if (index < 0 || this.head === null)
            return false;

        if (index === 0) {
            this.head = this.head.next;
            return true;
        }

        let current = this.head;
        let count = 0;
        while (current !== null && count < index - 1) {
            current = current.next;
            count++;
        }

        if (current === null || current.next === null)
            return false;

        current.next = current.next.next;
        return true;
    }

    /**
     * @return {number[]}
     */
    getValues() {
        let values = [];
        let current = this.head;
        while (current !== null) {
            values.push(current.data);
            current = current.next;
        }
        return values;
    }
}

// Example usage:
let list = new LinkedList();
list.insertHead(1);
list.insertTail(2);
list.insertHead(0);
list.remove(1);
console.log(list.getValues()); // Output: [0, 2]
